src/SignalR/common/SignalR.Common/src/Protocol/HandshakeProtocol.cs

```diff
@@ -192,63 +167,54 @@ public static bool TryParseRequestMessage(ref ReadOnlySequence<byte> buffer, out
                 return false;
             }
 
-            var textReader = Utf8BufferTextReader.Get(payload);
-            try
+            var reader = new Utf8JsonReader(in payload, isFinalBlock: true, state: default);
+
+            reader.CheckRead();
+            reader.EnsureObjectStart();
+
+            string protocol = null;
+            int? protocolVersion = null;
+
+            while (reader.CheckRead())
             {
-                using (var reader = JsonUtils.CreateJsonTextReader(textReader))
+                if (reader.TokenType == JsonTokenType.PropertyName)
                 {
-                    JsonUtils.CheckRead(reader);
-                    JsonUtils.EnsureObjectStart(reader);
-
-                    string protocol = null;
-                    int? protocolVersion = null;
+                    var memberName = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;
```

Comment:
@ahsonkhan Have you considered having a helper method for comparing the current value that worries about this for you? It would avoid developers making the mistake of not using ValueSequence, and could do something more efficient than allocating an array when doing the compare.

Reply:
https://github.com/dotnet/corefx/issues/34938

Reply:
I initially only had `ValueSpan` and would call `ToArray` on the ReadOnlySequence slice which circumvents this usability issue. However, it results in unnecessary allocations for JSON elements the user just wants to read past and not materialize/look into. So, we decided to leave the allocation to the caller since they have the context on when they actually care about the value or not.

> it could do something more efficient than allocating an array when doing the compare.

We could consider providing such a helper for compares. That is interesting feedback :)

The concern I have is, if the compare returns true, the user would have to call ToArray on the ReadOnlySequence anyway, so now we walk the sequence twice (once for the helper to check/compare, and if true, the caller materializing the value to some string/byte[]). We could save on allocations (by potentially copying to an array pool or some custom EqualToSequence method), but I don't know if it helps performance. How often is the search true versus false?

Reply:
Before:

```cs
var memberName = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;

if (memberName.SequenceEqual(ProtocolPropertyNameBytes))
{
    protocol = reader.ReadAsString(ProtocolPropertyNameBytes);
}
else if (memberName.SequenceEqual(ProtocolVersionPropertyNameBytes))
{
    protocolVersion = reader.ReadAsInt32(ProtocolVersionPropertyNameBytes);
}
else
{
    reader.Skip();
}
```

After:

```cs
if (reader.ValueEquals(ProtocolPropertyNameBytes))
{
    protocol = reader.ReadAsString(ProtocolPropertyNameBytes);
}
else if (reader.ValueEquals(ProtocolVersionPropertyNameBytes))
{
    protocolVersion = reader.ReadAsInt32(ProtocolVersionPropertyNameBytes);
}
else
{
    reader.Skip();
}
```

Reply:
I did this manually with a helper function for `ReadOnlySequence.SequenceEqual` and it helped the multi-segment case but slowed the single-segment case

Reply:
So you've already written `ReadOnlySequence.SequenceEqual` and calling that only when HasValueSequence is false slows down the single-segment case significantly? If you were to do something like the following, I hope the single-segment case wouldn't too much of a hit:

```c#
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static bool ValueEquals(in Utf8JsonReader reader, Span<byte> value)
{
    return reader.HasValueSequence ?
        reader.ValueSequence.SequenceEqual(value) :
        reader.ValueSpan.SequenceEqual(value);
}
```

Reply:
I give `ValueEquals` a try, maybe I was doing something dumb.

Reply:
IMO don't worry about optimizing this in SignalR. These messages are always going to be in one segment. I think this is an improvement for Utf8JsonReader.

Reply:
I agree that it's not important for this change. I'm just curious why @BrennanConroy's helper function is slower. I'm thinking maybe it's the Utf8JsonReader parameter getting copied.